File: leetcode/2466. Count Ways To Build Good Strings.py
class Solution:
    def f(self, high: int, zero: int, one:int) -> int:
        if self.dp[high] != -1: return self.dp[high]
        c = 0
        if high >= zero:
            c += self.f(high-zero, zero, one)
        if high >= one:
            c += self.f(high-one, zero, one)
        self.dp[high] = c % self.mod
        return self.dp[high]

    def countGoodStrings(self, low: int, high: int, zero: int, one: int) -> int:
        self.dp = [1] + [-1] * high
        self.mod = 1e9+7
        ans = 0
        for i in range(high, low-1, -1):
            ans += self.f(i, zero, one)
        return int(ans%self.mod)

File: leetcode/2466. Count Ways To Build Good Strings.cpp
class Solution {
public:

    vector<long long> dp;
    const int mod = 1e9+7;
    int _zero, _one;

    long long append(long long l){
        if (dp[l] != -1) return dp[l];
        long long ans = 0;
        if (l >= _zero){
            ans += append(l-_zero);
        }
        if (l >= _one){
            ans += append(l-_one);
        }
        if (ans >= mod) ans %= mod;
        dp[l] = ans;
        return ans;
    }


    int countGoodStrings(int low, int high, int zero, int one) {
        _zero = zero;
        _one = one;
        long long ans = 0;
        dp.resize(high+1, -1);
        dp[0] = 1;
        for (int i=high; i>=low; i--){
            ans += append(i);
        }
        return ans%mod;
    }
};